docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,551 @@
+   concatenation of `config.injection`, with the in-package resource kind and
+   name, and the value set to `False`. Note that since the package author
+   controls the name of the resource, kind and name are sufficient to
+   disambiguate the injection point. We will call this ConditionType the
+   "injection point ConditionType".
+1. For each required injection point, the injection point ConditionType will be
+   added to the PackageRevision `readinessGates`. Optional injection points'
+   ConditionTypes must not be added to the `readinessGates` by the
+   PackageVariant controller, but humans or other actors may do so at a later
+   date, and the package variant controller should not remove them on subsequent
+   reconciliations. Also, this relies upon `readinessGates` gating publishing
+   the package to a *deployment* repository, but not gating publishing to a
+   blueprint repository.
+1. The injection processing will proceed as follows. For each injection point:
+   - If the resource schema of the injection point is not available in the
+     cluster, then the injection point ConditionType will be set to `False`,
+     with a message indicating that the schema is missing, and processing should
+     proceed to the next injection point. Note that for `optional` injection
+     points, not having the schema may be intentional and not an error.
+   - If the resource schema of the injection point does not contain a `spec`
+     field, then the injection point ConditionType will be set to `False`, with
+     a message explaining the error, and processing should proceed to the next
+     injection point.
+   - The controller will now identify all in-cluster objects in the same
+     namespace as the PackageVariant resource, with GVK matching the injection
+     point (the in-package resource).
+   - The controller will look through the list of injection selectors in
+     order and checking if any of the in-cluster objects match the selector. If
+     so, that in-cluster object is selected, and processing of the list of
+     injection selectors stops. Note that in the initial version, the selectors
+     are only by name; thus, at most one match is possible for any given
+     selector.
+   - If no in-cluster object is selected, the injection point ConditionType will
+     be set to `False` with a message that no matching in-cluster resource was
+     found, and processing proceeds to the next injection point.
+   - If a matching in-cluster object is selected, then it is injected as
+     follows:
+     - The `spec` field from the in-cluster resource is copied to the `spec`
+       field of the in-package resource (the injection point), overwriting it.
+     - An annotation with name `kpt.dev/injected-resource-name` and value set to
+       the name of the in-cluster resource is added (or overwritten) in the
+       in-package resource.
+1. The PackageVariant resource must have a condition of type `ConfigInjected`.
```

Comment:
The language here doesn't make it clear if the package must always have this condition set, or if it is added by automation. It might be useful to clarify it.